Cl 2326 fe unit integration tests for layout settings

brentguf opened the pull request:
## Checklist

- [ ] Added entry to changelog
<details>
<summary>More info</summary>
Add a concise line to the 'Next release' section of the changelog (docs/README.md) so people other than developers can understand what has changed where. E.g. 'Added an error message to the project name field of the project edit form (Admin > Projects > Edit)'.
</details>

- [ ] WCAG 2.1 AA proof
<details>
<summary>More info</summary>
For front-end devs only. Is your work conforming with the WCAG 2.1 AA rules? If you need more info, read the [a11y page](https://www.notion.so/citizenlab/a11y-7568f83d42ab4895ac133b89d358997b) on our Notion.
</details>

- [ ] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [ ] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

## Links

- [citizenlab-ee PR](**put URL here** or remove)
- [Specs](**put URL here** or remove)
- [Epic Deployment](**put URL here** or remove)

## How urgent is a code review?

Let the reviewer(s) know how urgent the code review is, so they can prioritize their work accordingly. Be specific (e.g. by Wednesday, end of the day/this week/... is better than 'urgent' or 'very urgent'). Optionally provide a word of explanation on your deadline.

brentguf committed:
BannerImageFields tests with fixed-ratio layout

brentguf committed:
Merge branch 'CL-2215-fe-adding-toggle-to-banner-form-to-hide-for-signed-in-users' of https://github.com/CitizenLabDotCo/citizenlab into CL-2326-fe-unit-integration-tests-for-layout-settings

brentguf committed:
1 Layout preview selector describe block

brentguf committed:
Remove console.log

brentguf committed:
Merge branch 'CL-1849/image-dimensions' of https://github.com/CitizenLabDotCo/citizenlab into CL-2326-fe-unit-integration-tests-for-layout-settings

brentguf committed:
Remove type casting from test

brentguf committed:
Remove unnecessary condition

brentguf committed:
Improve Layout preview selector

brentguf committed:
Update snapshots

brentguf committed:
Restructure BannerImageFields tests

brentguf committed:
Merge branch 'CL-1849/image-dimensions' of https://github.com/CitizenLabDotCo/citizenlab into CL-2326-fe-unit-integration-tests-for-layout-settings

brentguf committed:
WIP

brentguf committed:
Image cropper fixed ratio layout test

brentguf committed:
Improve Image cropper tests

brentguf committed:
Image overlay controls WIP

brentguf committed:
Simplify Image overlay controls tests

brentguf committed:
Simplify Layout preview selector tests

brentguf committed:
Merge branch 'CL-1849/image-dimensions' of https://github.com/CitizenLabDotCo/citizenlab into CL-2326-fe-unit-integration-tests-for-layout-settings

brentguf committed:
Update component library version

brentguf committed:
Various fixes

brentguf committed:
Resolve useEffect memory leak warning

brentguf committed:
Move convertHeaderToUploadFile out of useEffect

brentguf committed:
Move isUploadFile to fileUtils

brentguf committed:
Remove non-sensical array around header image upload files

brentguf committed:
BannerImageFields useEffect improvement WIP

brentguf committed:
Fix image cropper display

brentguf committed:
Remove comments

brentguf committed:
Test overlay settings presence

brentguf committed:
Move setBannerError to image upload

brentguf committed:
Remove failing 'saved image' tests

brentguf committed:
Merge branch 'master' of https://github.com/CitizenLabDotCo/citizenlab into CL-2326-fe-unit-integration-tests-for-layout-settings

brentguf committed:
Lint errors

brentguf committed:
Unused imports

brentguf committed:
Unused props

brentguf committed:
Remove non-useful tests

brentguf committed:
Improve imageIsSaved condition

brentguf committed:
Improved conditions overview for banner image fields

brentguf committed:
Remove impossible types for banner_layout

brentguf committed:
Add types for showConditions

brentguf committed:
Undo changes to useEffect

brentguf commented on front/app/containers/Admin/pagesAndMenu/EditHomepage/HeroBanner/index.tsx:
```diff
@@ -192,7 +192,6 @@ const EditHomepageHeroBannerForm = () => {
                 localSettings.banner_signed_out_header_overlay_opacity
               }
               headerBg={localSettings.header_bg}
-              setFormStatus={setFormStatus}
```
Was not really used, only as a dependency in useEffect, but it was not used in useEffect.

brentguf committed:
Comments

brentguf commented on front/app/containers/Admin/pagesAndMenu/containers/GenericHeroBannerForm/BannerImageFields/index.tsx:
```diff
@@ -59,31 +64,18 @@ const BannerImageField = ({
-    // to a format that can be displayed. this is done locally
-    // when the image is changed but needs to be done manually
-    // to process the initial API response
-    const convertHeaderToUploadFile = async (fileInfo) => {
-      if (fileInfo) {
-        const tenantHeaderBg = await convertUrlToUploadFile(fileInfo);
-        const headerBgUploadFile = !isNilOrError(tenantHeaderBg)
-          ? [tenantHeaderBg]
-          : [];
-        setHeaderLocalDisplayImage(headerBgUploadFile);
-        setBannerError(null);
-      }
```
Only moved outside of the useEffect.

brentguf commented on front/app/containers/Admin/pagesAndMenu/containers/GenericHeroBannerForm/BannerImageFields/index.tsx:
```diff
@@ -94,37 +86,70 @@ const BannerImageField = ({
+  const imageShouldBeSaved = headerLocalDisplayImage
+    ? headerLocalDisplayImage.remote
+    : false;
+
```
Replaces `!isImageSaved` as per Alex' suggestion.

brentguf committed:
conditions include custom page banner layout

brentguf commented on front/app/containers/Admin/pagesAndMenu/containers/GenericHeroBannerForm/BannerImageFields/index.tsx:
```diff
@@ -94,37 +86,70 @@ const BannerImageField = ({
-    (bannerLayout === 'full_width_banner_layout' ||
-      (bannerLayout === 'fixed_ratio_layout' && imageIsSaved));
+    return conditions[bannerLayout];
+  };
```
Imo much clearer than the display conditions. Now we need to define it explicitly for each layout. Maybe this is almost as good as the unit/integration tests and could almost replace them. The unit/integration tests act as documentation/reference though.

brentguf commented on front/cypress/e2e/admin/pages_and_menu/update_homepage_hero_banner_content.cy.ts:
```diff
@@ -102,6 +102,7 @@ describe('Admin: update Hero Banner content', () => {
       'testimage.png'
     );
     cy.get('[data-cy="e2e-image-cropper"]'); // wait until image cropper is loaded and visible (image is cropped). The test is flaky without this line.
+    cy.get('#overlay-toggle').should('not.exist');
```
Added the overlay toggle checks in the E2E because it was very hard to test with React Testing Library.

alexander-cit replied:
Why did you use `id` and not `data-cy`?

brentguf replied:
Good question. Changing this.

brentguf commented on front/app/containers/Admin/pagesAndMenu/containers/GenericHeroBannerForm/HeaderImageDropzone.tsx:
```diff
@@ -36,7 +36,7 @@ interface Props {
   onAdd: (newImage: UploadFile[]) => void;
   onRemove: () => void;
   headerError: string | null;
-  header_bg: UploadFile[] | null;
```
Didn't understand why this needs to be an array if we can only add 1 file at most.

brentguf commented on front/app/containers/Admin/pagesAndMenu/containers/GenericHeroBannerForm/BannerImageFields/index.tsx:
```diff
@@ -59,31 +64,18 @@ const BannerImageField = ({
+
     convertHeaderToUploadFile(headerFileInfo);
   }, [headerBg]);
 
```
There is still a problem here. headerBg returns something different when we have an unsaved vs. saved image (base64 string vs. object of type ImageSizes) but the types for header_bg in homepageSettings and customPages are not making this obvious (because they are wrong/incomplete). I'll address it in the FE CDM.

alexander-cit commented on front/app/containers/Admin/pagesAndMenu/containers/GenericHeroBannerForm/BannerImageFields/BannerImageFields.test.tsx:
```diff
@@ -1,161 +1,192 @@
-} as Props;
+};
 
-describe('BannerImageFields', () => {
```
I'm not sure this convention is strict, but in Rails and in some jest examples, they still group all the tests for one function (component) into one `describe` https://jestjs.io/docs/api#describename-fn#:~:text=You%20can%20also%20nest%20describe%20blocks

But of course up to you.

brentguf replied:
Done ✅

alexander-cit commented on front/app/containers/Admin/pagesAndMenu/containers/GenericHeroBannerForm/BannerImageFields/BannerImageFields.test.tsx:
```diff
@@ -1,161 +1,192 @@
-    const file = new File(['file'], 'file.png', {
-      type: 'image/png',
-    });
+  describe('when layout is fixed-ratio layout', () => {
```
The behavior of full-width and fixed-ratio is the same regarding "Enable overlay." Why are the tests different?

brentguf replied:
For the fixed-ratio banner, you can only turn the overlay on after the image is saved because we can't show the overlay on top of the cropper.

alexander-cit reviewed (approved):
Just a few minor comments that can be basically ignored.

brentguf commented on front/app/containers/Admin/pagesAndMenu/containers/GenericHeroBannerForm/BannerImageFields/index.tsx:
```diff
@@ -59,72 +67,95 @@ const BannerImageField = ({
-  const imageIsSaved = headerLocalDisplayImage?.[0].remote || false;
   const hasLocalHeaderImage = !isNilOrError(headerLocalDisplayImage);
+  const imageShouldBeSaved = headerLocalDisplayImage
+    ? headerLocalDisplayImage.remote
```
This needs to be `!headerLocalDisplayImage.remote` instead.

brentguf replied:
My tests didn't fail even though this value was wrong. 🤔

brentguf replied:
@IvaKop I guess it's better to replace all tests that involve "saved images" with E2E tests. The tests here don't seem reliable for that. Wdyt?

alexander-cit replied:
[This test](https://github.com/CitizenLabDotCo/citizenlab/blob/3bc53601cec2dc100595ea4e91a3859e60f2ee63/front/app/components/admin/ProjectableHeaderBgUploader/ProjectableHeaderBgUploader.test.tsx) detects the same bug [in this line](https://github.com/CitizenLabDotCo/citizenlab/blob/3bc53601cec2dc100595ea4e91a3859e60f2ee63/front/app/components/admin/ProjectableHeaderBgUploader/index.tsx#L43). So, I think a unit test is still fine here.

IvaKop replied:
If it works, I would say let's keep it? No harm in having the functionality covered by both unit and e2e tests (assuming we cover this with Cypress, already).

brentguf replied:
I'm confused for a moment. Will keep you updated when it's clear.

alexander-cit replied:
You didn't ask me, but I'll reply! :D 

In my understanding, `not.toBeInTheDocument` is not reliable because it's always `true` at the beginning. So, you should either wait for some time or to use the positive check `toBeInTheDocument` to make it reliable.

alexander-cit replied:
As it's done here for example https://github.com/CitizenLabDotCo/citizenlab/blob/3bc53601cec2dc100595ea4e91a3859e60f2ee63/front/app/components/admin/ProjectableHeaderBgUploader/ProjectableHeaderBgUploader.test.tsx#L34

brentguf replied:
Hang on for a moment. There are bugs in my code, so I need to fix these first. Wrong elements are displaying.

I'll ask you again to review when I'm done :D

brentguf committed:
Wrap with parent describe

brentguf committed:
Use data-cy > id

brentguf committed:
Merge branch 'master' of https://github.com/CitizenLabDotCo/citizenlab into CL-2326-fe-unit-integration-tests-for-layout-settings

brentguf committed:
headerBg fixes

brentguf committed:
imageShouldBeSaved correct value

IvaKop commented on front/app/containers/Admin/pagesAndMenu/containers/GenericHeroBannerForm/BannerImageFields/BannerImageFields.test.tsx:
```diff
@@ -1,161 +1,204 @@
   });
 });
+
+async function uploadLocalImageForHeroBanner() {
```
I would put this on the top of the file, it's a bit hard to find down here.

IvaKop commented on front/app/containers/Admin/pagesAndMenu/containers/GenericHeroBannerForm/BannerImageFields/index.tsx:
```diff
@@ -56,72 +64,95 @@ const BannerImageField = ({
+    ? !headerLocalDisplayImage.remote
+    : false;
+
+  const showConditions = (bannerLayout: THomepageBannerLayout) => {
```
Nice! Very clear what is happening here.

IvaKop reviewed (approved):
LGTM! Nothing major to point out.

brentguf committed:
headerBg is not an array

brentguf committed:
Overlay element shouldn't show for certain layouts

brentguf committed:
Fixed-ratio: overlay should only show for saved images

brentguf committed:
Add data-cy to correct element (fix E2E)

brentguf committed:
Remove unneeded banner_layout fallbacks

brentguf committed:
Remove tests for saved image

brentguf committed:
Remove saved headerBg from props

brentguf committed:
Improve comments

brentguf committed:
Remove overlay on image deletion

brentguf committed:
Cover some banner image test cases that require saving of the image.

brentguf committed:
Test overlay toggle for full width banner layout

brentguf commented on front/app/containers/Admin/pagesAndMenu/EditHomepage/HeroBanner/index.tsx:
```diff
@@ -145,6 +145,7 @@ const EditHomepageHeroBannerForm = () => {
 
   const handleOnBannerImageRemove = () => {
     handleOnChange('header_bg', null);
+    handleOnOverlayChange(null, null);
```
If you remove an image now, the overlay will also be turned off. Thought that was desirable behaviour while I was trying it.

brentguf commented on front/app/containers/Admin/pagesAndMenu/containers/GenericHeroBannerForm/HeaderImageDropzone.tsx:
```diff
@@ -96,15 +99,24 @@ const HeaderImageDropzone = ({
   const previewOverlayElement =
-    //  We check for typeof of opacity because 0 would coerce to false.
-    isBannerWithOverlay &&
+    showPreviewOverlayForLayout(layout) &&
```
Wrongly removed this, because we're currently not removing the overlay when we change the layout. This means it was showing for two-row layout now.

brentguf committed:
Comment onRemove type

brentguf committed:
setBannerError in image removal function

brentguf committed:
Prevent unnecessary convertHeaderToUploadFile calls

brentguf committed:
Revert DropzoneLabel to span (file picker bug)

brentguf commented:
@alexander-cit @IvaKop Hopefully the final review. Since the last review I:
- Deleted untrustworthy integration tests depending on saved images having loaded async & replaced these with E2E tests
- Fixed 2 bugs that I had introduced in the meanwhile
- Some general further code improvements

Note I also had to revert back the change of the dropzone label to a span because it caused buggy behaviour with the file picker unfortunately. It wasn't straightforward to fix.

The failing `front/cypress/e2e/admin/pages_and_menu/update_homepage_hero_banner_content.cy.ts` on master passes here as well.

IvaKop reviewed (approved):
Nice work 🚀

alexander-cit reviewed (approved):
I still think it was possible to test it properly with a unit test (as in the example I mentioned previously). 

E2E tests are not free. They are flaky, slow, and in general hard to support (in fact, they've been currently failing for more than a week which makes them useless now).

But it's not critical.

brentguf commented:
> I still think it was possible to test it properly with a unit test (as in the example I mentioned previously).
> 
> E2E tests are not free. They are flaky, slow, and in general hard to support (in fact, they've been currently failing for more than a week which makes them useless now).
> 
> But it's not critical.

@alexander-cit Sorry, forgot about your proposal. But I really want to wrap up this task so I'm not going to change it now. Thanks for your help though.

alexander-cit commented:
@brentguf That's totally fine!
